### PR TITLE
Support Network Policy for UDN namespaces

### PR DESCRIPTION
**Pull request by pperiyasamy**
#### What this PR does and why is it needed

This PR has required changes to support Network Policy on User Defined Network.

- [X] NetPol Support on L3 UDN network.
- [X] NetPol support on on L2 UDN network.

TODOs:

- [ ] Run NetPol upstream conformance tests over UDN network.
- [ ] Add hairpin allow ACLs for primary UDNs network policies (https://issues.redhat.com/browse/SDN-5290).

#### Which issue(s) this PR fixes

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Comment by npinaeva** [comment]
there are some netpol-related test failures. Also, could you please add a couple of e2es for the new netpol behaviour?

**Comment by pperiyasamy** [comment]
hit this ci flake https://github.com/ovn-org/ovn-kubernetes/issues/4299 in this run:
https://github.com/ovn-org/ovn-kubernetes/actions/runs/10716745664/job/29716177061?pr=4690

File: go-controller/pkg/util/pod_annotation.go
Comment (jcaamano) [inline go-controller/pkg/util/pod_annotation.go:328]: Please note that `nInfo.IsSecondary` return `true` for primary networks as well.

However, how is this implementation any different functionally than `SecondaryNetworkPodIPs`?

File: go-controller/pkg/ovn/base_network_controller.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller.go:704]: Why do we need to do this?

Isn't this network controller the one for the active network for the namespace?

In which cases do we expect `namespacePrimaryNetwork` to be different than `bnc.NetInfo`?

Comment (jcaamano) [reply go-controller/pkg/ovn/base_network_controller.go:704]: Comments always help though

File: go-controller/pkg/ovn/base_network_controller_policy.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller_policy.go:230]: nit: `hairpin`

File: go-controller/pkg/ovn/base_network_controller_policy.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller_policy.go:228]: Why don't we need the hairpin allow ACLs for primary networks? Add comment.

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller_secondary.go:83]: I guess these could live in the base event handler if they are going to be handled equally for all network types.

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller_secondary.go:770]: I guess througout the PR, why do we check `IsNetworkSegmentationSupportEnabled` along with `IsPrimaryNetwork`?

When don't do the equivalent with secondary networks checking `MultiNetworkEnabled` along with `IsSecondary()`

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller_secondary.go:777]: Don't we need a different handler for NetPol vs MultiNetPol?

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller_secondary.go:786]: I guess this can go to the base_network_controller

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (jcaamano) [reply go-controller/pkg/ovn/base_network_controller_secondary.go:770]: Also, multinetpol should only apply to non primary networks?

File: go-controller/pkg/ovn/base_secondary_layer2_network_controller.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_secondary_layer2_network_controller.go:37]: nit: just checking `oc.policyHandler != nil` should be enough. As already mentioned, you need a different handler for multinetpols

File: go-controller/pkg/ovn/secondary_layer2_network_controller.go
Comment (jcaamano) [inline go-controller/pkg/ovn/secondary_layer2_network_controller.go:66]: all this could go into base_event_handler

Comment (jcaamano) [reply go-controller/pkg/ovn/secondary_layer2_network_controller.go:66]: Also, a bit below, why for RecordUpdateEvent we are calling `h.baseHandler.recordAddEvent` instead of `h.baseHandler.recordUpdateEvent`? 

PTAL @tssurya for awareness

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (jcaamano) [reply go-controller/pkg/ovn/base_network_controller_secondary.go:777]: I see that you initialize them different. However, I would still use separate fields for clarity.

File: go-controller/pkg/ovn/secondary_layer2_network_controller.go
Comment (tssurya) [reply go-controller/pkg/ovn/secondary_layer2_network_controller.go]: yeah sounds like a copy paste snafu nice catch, should be add, update, delete respectively. I'll take care of a cleanup PR for that

File: go-controller/pkg/ovn/base_network_controller.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller.go:708]: added comment to explain this.

File: go-controller/pkg/ovn/base_network_controller_policy.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller_policy.go:231]: added comment about why hairpin allow ACLs disabled now for UDN. will revisit once we have cluster port group exists per network. currently this is not impacting unit tests and e2e's.

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller_secondary.go:83]: all the secondary networks event handling done here, similarly for default network, it's being handled in default network controller.  do you mean to say PolicyType handling can be moved into BaseNetworkController as policy event handling is common for both ?

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller_secondary.go:772]: yes, done.

File: go-controller/pkg/util/pod_annotation.go
Comment (pperiyasamy) [reply go-controller/pkg/util/pod_annotation.go]: > Please note that `nInfo.IsSecondary` return `true` for primary networks as well.

yes @jcaamano , thanks for the catch, it was actually causing failures for upstream netpol tests. fixed it now.

> 
> However, how is this implementation any different functionally than `SecondaryNetworkPodIPs`?

The retrieval of IPs from `SecondaryNetworkPodIPs` is retrieving nad name based on `k8s.v1.cni.cncf.io/networks` annotation which doesn't work for primary networks. hence separate handling is needed for primary networks.

File: go-controller/pkg/ovn/base_network_controller_namespace.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller_namespace.go]: with two handlers, this check is not needed anymore.

File: go-controller/pkg/ovn/base_network_controller_policy.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller_policy.go]: done.

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller_secondary.go]: yes, i think multinetpol is applicable only for non primary networks.


> I guess througout the PR, why do we check `IsNetworkSegmentationSupportEnabled` along with `IsPrimaryNetwork`?
> 
This is a common practice being used all over the places. isn't this kind of gating check ? 
If this is not needed at secondary network controller level, can remove this.

File: go-controller/pkg/ovn/base_secondary_layer2_network_controller.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_secondary_layer2_network_controller.go]: yes, done.

File: go-controller/pkg/ovn/secondary_layer2_network_controller.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/secondary_layer2_network_controller.go]: @tssurya I've addressed it with this PR, please take a look.

File: go-controller/pkg/util/pod_annotation.go
Comment (jcaamano) [inline go-controller/pkg/util/pod_annotation.go:320]: Some things here:

- Un-marshaling the annotation for each NAD is inefficient
- Why would be a pod be annotated with a NAD of a different namespace?
- Use a well defined error. We should we using `newAnnotationNotSetError`, we are probably not doing it for historic reasons but I don't see why.

File: go-controller/pkg/ovn/base_network_controller.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller.go:708]: You could also comment here that `PodNadNames` is based on the network selection element so it wouldn't work. All this implementation could be in `util.PodNadNames` I think

File: test/scripts/e2e-cp.sh
Comment (jcaamano) [inline test/scripts/e2e-cp.sh:205]: Why do we need this when we already have 
https://github.com/ovn-org/ovn-kubernetes/blob/master/test/scripts/e2e-kind.sh#L28-L29

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (jcaamano) [reply go-controller/pkg/ovn/base_network_controller_secondary.go]: So if segmentation is not enabled, why woudl we have a network controller for a primary network? It shouldn't exist in the first place, already checked by the NAD controller. No harm I guess though.

File: go-controller/pkg/util/pod_annotation.go
Comment (jcaamano) [reply go-controller/pkg/util/pod_annotation.go]: Can you add a comment?

File: go-controller/pkg/ovn/base_network_controller.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller.go:712]: added comment and i tried to move this logic into util.PodNadNames, but it requires nadLister nadlister.NetworkAttachmentDefinitionLister, so needed intrusive changes across all the calling methods. do you still this to be part of this PR or can it be done separately ?

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller_secondary.go:729]: oh! no problem, didn't notice it, using different variables now.

File: test/scripts/e2e-cp.sh
Comment (pperiyasamy) [reply test/scripts/e2e-cp.sh:205]: the e2e-kind.sh script does not run for network segmentation lane instead e2e-cp.sh script is [invoked](https://github.com/ovn-org/ovn-kubernetes/blob/master/.github/workflows/test.yml#L551) for it. hence another skip for named port netpol tests here.

File: go-controller/pkg/util/pod_annotation.go
Comment (pperiyasamy) [reply go-controller/pkg/util/pod_annotation.go]: > * Un-marshaling the annotation for each NAD is inefficient

do we have any other option other than parsing netwoks annotation ?

> * Why would be a pod be annotated with a NAD of a different namespace?

This function may be invoked with netInfo object which is global object for multiple namespaces, hence needed logic to fetch right NAD name for the namespace.

> * Use a well defined error. We should we using `newAnnotationNotSetError`, we are probably not doing it for historic reasons but I don't see why.

yes, using `newAnnotationNotSetError` for this purpose now, didn't see any issue during the tests.

File: go-controller/pkg/ovn/base_secondary_layer2_network_controller.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_secondary_layer2_network_controller.go]: sure, done.

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller_secondary.go]: done.

File: go-controller/pkg/ovn/base_network_controller_namespace.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller_namespace.go]: oh! yes, multinetpolicy is also enabled when multinetwork is set. 
removed  the check `!bnc.IsPrimaryNetwork() `now.

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller_secondary.go]: you are right, removed `IsNetworkSegmentationSupportEnabled` check whereever it's possible.

File: go-controller/pkg/util/pod_annotation.go
Comment (pperiyasamy) [reply go-controller/pkg/util/pod_annotation.go]: done.

File: go-controller/pkg/util/pod_annotation.go
Comment (jcaamano) [reply go-controller/pkg/util/pod_annotation.go]: > > * Un-marshaling the annotation for each NAD is inefficient
> 
> do we have any other option other than parsing netwoks annotation ?

You can parse once getting a map of NADs to IPs. Instead of parsing the full annotation for each NAD.

> > * Why would be a pod be annotated with a NAD of a different namespace?
> 
> This function may be invoked with netInfo object which is global object for multiple namespaces, hence needed logic to fetch right NAD name for the namespace.

So why not use `GetActiveNetworkForNamespace` which will give you the NADs for the namespace? This is related to the previous comment, you are parsing the annotation for each NAD when many of them might not even be relevant.

> 
> > * Use a well defined error. We should we using `newAnnotationNotSetError`, we are probably not doing it for historic reasons but I don't see why.
> 
> yes, using `newAnnotationNotSetError` for this purpose now, didn't see any issue during the tests.

Did you check that this is ok for previously existing uses of `UnmarshalPodAnnotation`? While it should be fine, we need to be careful.

Comment (jcaamano) [reply go-controller/pkg/util/pod_annotation.go]: > So why not use `GetActiveNetworkForNamespace` which will give you the NADs for the namespace? This is related to the previous comment, you are parsing the annotation for each NAD when many of them might not even be relevant.

Meaning, GetActiveNetworkForNamespace should give you a netinfo with a single NAD because there can only be one primary NAD per namespace, so the issue wouldn't be as important.

File: go-controller/pkg/ovn/base_network_controller.go
Comment (jcaamano) [reply go-controller/pkg/ovn/base_network_controller.go:712]: But the lister is already being passed to the global function `util.GetActiveNetworkForNamespace`

File: test/scripts/e2e-cp.sh
Comment (jcaamano) [reply test/scripts/e2e-cp.sh:205]: But is that is not running conformance tests? I am not aware that we have any `named port.+\[Feature:NetworkPolicy\]` in our control plane tests but I might be missing something.

File: go-controller/pkg/util/pod_annotation.go
Comment (jcaamano) [inline go-controller/pkg/util/pod_annotation.go:365]: So `nadName` is a namespaced name: `<namespace>/<name>`. 

So you can just parse the namespace and match it against the pod namespace.

I don't think we need to check the role. If we have the pod annotated with a non-primary NAD that we have as primary in NetInfo, we have bigger problems.

Comment (jcaamano) [reply go-controller/pkg/util/pod_annotation.go:365]: I have asked @trozet to add a `GetNADs(namespace string)` method to avoid having to parse the namespaced name however.

File: go-controller/pkg/util/pod_annotation.go
Comment (jcaamano) [inline go-controller/pkg/util/pod_annotation.go:363]: `!strings.HasPrefix(nadName, namespace+'/')`

Being strict, we should be using `SplitMetaNamespaceKey` however.

Comment (pperiyasamy) [reply go-controller/pkg/util/pod_annotation.go]: ok, using `SplitMetaNamespaceKey`  function to check namespace now.

File: go-controller/pkg/ovn/base_network_controller.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller.go:952]: I don't think update for network policies is going to be called, from 
https://github.com/jcaamano/ovn-kubernetes/blob/14afcbf9621264f6d1de32107bfdcf77f655bbbc/go-controller/pkg/ovn/base_event_handler.go#L25-L26

File: go-controller/pkg/ovn/base_network_controller.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller.go:922]: Technically I would just log here because the retries would be useless

File: test/scripts/e2e-cp.sh
Comment (jcaamano) [reply test/scripts/e2e-cp.sh]: We discussed not adding this in this PR for the time being.

File: go-controller/pkg/ovn/base_network_controller_policy.go
Comment (jcaamano) [reply go-controller/pkg/ovn/base_network_controller_policy.go:231]: https://issues.redhat.com/browse/SDN-5290

File: go-controller/pkg/ovn/base_network_controller.go
Comment (pperiyasamy) [reply go-controller/pkg/ovn/base_network_controller.go]: right, removed the method as it's no longer needed.

**Comment by jcaamano** [comment]
Potential flake
 [FAIL] Network Segmentation: services on a user defined primary network should be reachable through their cluster IP and node port [It] L3 dualstack primary UDN, cluster-networked pods, NodePort service
 https://github.com/ovn-org/ovn-kubernetes/actions/runs/10900030212/job/30253567040?pr=4690

**Comment by jcaamano** [comment]
> Potential flake [FAIL] Network Segmentation: services on a user defined primary network should be reachable through their cluster IP and node port [It] L3 dualstack primary UDN, cluster-networked pods, NodePort service https://github.com/ovn-org/ovn-kubernetes/actions/runs/10900030212/job/30253567040?pr=4690

@pperiyasamy does not look like a flake

**Review by trozet — APPROVED** [review]
minor comment nits, lgtm

File: test/e2e/network_segmentation_policy.go
Comment (trozet) [inline test/e2e/network_segmentation_policy.go:125]: the title of this PR says netpol for L3, but here are L2 test cases and changes to layer 2 controller. So is L2 also supported and the title just needs to be updated?

Comment (pperiyasamy) [reply test/e2e/network_segmentation_policy.go:127]: yes @trozet , updated title and description accordingly now.

File: go-controller/pkg/util/pod_annotation.go
Comment (jcaamano) [inline go-controller/pkg/util/pod_annotation.go:391]: This is based on two assumptions:

- There is a single NAD per namespace for primary networks. However, there is nothing currently that prevents this from happening.
- We are aware of the aware of the primary network of a pod. However this might be reacey.

The first issue should be handled in some way by the NAD controller. It might do that through `GetActiveNetworkForNamespace`. For the second issue, `GetActiveNetworkForNamespace` offers a bit more of protection since it will check against the UDN CRDs as well.

I don't want to hold this because of that. But I am entertaining the idea that we could just implement `GetActiveNetworkForNamespace` in NetInfo that would just end up calling the nad controller `GetActiveNetworkForNamespace` which is feasible since the nad controller is the one building those NetInfo after all.

File: test/e2e/network_segmentation_policy.go
Comment (jcaamano) [inline test/e2e/network_segmentation_policy.go:205]: Don't use defer to clean up please, it hinders troubleshooting. Use `f.AddNamespacesToDelete`; or manually in `AfterEach` depending on how the framework is configured and if the test case failed or not.

**Comment by npinaeva** [comment]
Is it intended to configure netpol for both default and primary UDN at the same time?

**Comment by pperiyasamy** [comment]
> Is it intended to configure netpol for both default and primary UDN at the same time?

yes @npinaeva , The netpol for default and primary UDN networks would coexist with this PR as well. The existing shard-conformance CI lane tests netpol over default network and we had few e2e's to test NetPol over UDN network. 
We also ran upstream NetPol tests over UDN L3 network on this PR and it worked fine. We're waiting for https://github.com/ovn-org/ovn-kubernetes/pull/4653 to run upstream tests over L2 network. Does it suffice ? or do you want to cover something more ?

File: test/e2e/network_segmentation_policy.go
Comment (pperiyasamy) [reply test/e2e/network_segmentation_policy.go]: yes @jcaamano , modified the test to use `f.AddNamespacesToDelete`.

File: go-controller/pkg/util/pod_annotation.go
Comment (pperiyasamy) [reply go-controller/pkg/util/pod_annotation.go:391]: I like the idea of having `GetActiveNetworkForNamespace` inside netInfo.

**Comment by npinaeva** [comment]
> yes @npinaeva , The netpol for default and primary UDN networks would coexist with this PR as well. The existing shard-conformance CI lane tests netpol over default network and we had few e2e's to test NetPol over UDN network. We also ran upstream NetPol tests over UDN L3 network on this PR and it worked fine. We're waiting for #4653 to run upstream tests over L2 network. Does it suffice ? or do you want to cover something more ?

So when primary UDN exists, we will configure network policy both for UDN and for default network, which means 2 times more port groups, acls, address sets, which will not be used (as default network traffic has its own ACLs that deny almost everything already). So UDN netpol will work, but not sure if it is the most optimal solution.

**Comment by jcaamano** [comment]
> So when primary UDN exists, we will configure network policy both for UDN and for default network, which means 2 times more port groups, acls, address sets, which will not be used (as default network traffic has its own ACLs that deny almost everything already). So UDN netpol will work, but not sure if it is the most optimal solution.

@npinaeva has a point.

So when handling a policy we would have to `GetActiveNetworkForNamespace(netpolNamespace)` and check that it matches our network. Seems straight forward enough.

**Comment by pperiyasamy** [comment]
> @npinaeva has a point.
> 
> So when handling a policy we would have to `GetActiveNetworkForNamespace(netpolNamespace)` and check that it matches our network. Seems straight forward enough.

Good find ! addressed this with commit [b5d3dc4](https://github.com/ovn-org/ovn-kubernetes/pull/4690/commits/b5d3dc4ced6219d35c85e339e12ac3c7e9327ebc).

File: go-controller/pkg/ovn/default_network_controller.go
Comment (jcaamano) [inline go-controller/pkg/ovn/default_network_controller.go:742]: Why don't you do this in `base_network_controller.go`?

You would just have to check `bnc.GetNetwokrName() != netinfo.GetNetworkName()`

How would you otherwise prevent the netpol that is processed by the default network controller to be processed as well on primary network controllers?

Comment (pperiyasamy) [reply go-controller/pkg/ovn/default_network_controller.go]: ok @jcaamano , having this in  `base_network_controller.go`  check is a good idea. done.
one question though: secondary network controllers never handle default network event ? correct ?

Comment (jcaamano) [reply go-controller/pkg/ovn/default_network_controller.go]: No sure what you mean. Non primary but still secondary networks don't handle events for `factory.PolicyType` because the retry framework should not be setup for them.

**Review by jcaamano — APPROVED** [review]

**Comment by tssurya** [comment]
wohoo :) thanks @jcaamano / @pperiyasamy for working on this and getting it to a good shape

**Pull request merged**
